### PR TITLE
feat(sso): Allow reading SSO clientID from a secret.

### PR DESCRIPTION
**Pull request by vladlosev**
This change adds support for reading the SSO client ID value from a secret, just like the client secret is read already. That will make it easier to provide that the client ID value externally, via facilities like external secrets, which in turn makes rotating secrets much easier.

Besides adding unit tests, tested this by running `make start AUTH_MODE=sso` and verifying the login process with dex and its sample app.

Checklist:

* [ ] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this is a chore.
* [x] The title of the PR is (a) [conventional](https://www.conventionalcommits.org/en/v1.0.0/), (b) states what changed, and (c) suffixes the related issues number. E.g. `"fix(controller): Updates such and such. Fixes #1234"`.  
* [x] I've signed the CLA.
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My builds are green. Try syncing with master if they are not. 
* [x] My organization is added to [USERS.md](https://github.com/argoproj/argo/blob/master/USERS.md).

### CODE REVIEW

File: server/auth/sso/sso.go [hunk 1]
```diff
@@ -39,18 +40,50 @@ type sso struct {
 }
 
 type Config struct {
-	Issuer       string                  `json:"issuer"`
-	ClientID     string                  `json:"clientId"`
-	ClientSecret apiv1.SecretKeySelector `json:"clientSecret"`
-	RedirectURL  string                  `json:"redirectUrl"`
+	Issuer         string                   `json:"issuer"`
+	ClientID       string                   `json:"clientId"`
```
[the remark this answers is not in the document]
Comment (vladlosev) [reply server/auth/sso/sso.go]: I've updated the code to make `ClientID` a `SecretKeySelector `. The configuration is simpler this way, for sure! Manifests were also updated to match the change, including `docs/workflow-controller-configmap.yaml`.

Testing the SSO mode with `make start` worked as well: the login page redirected to Dex and I was able to authorize using the sample app.

File: go.mod [hunk 2]
```diff
@@ -49,6 +49,7 @@ require (
 	github.com/lib/pq v1.3.0 // indirect
 	github.com/mailru/easyjson v0.7.1 // indirect
 	github.com/mattn/go-colorable v0.1.4 // indirect
+	github.com/mattn/goreman v0.3.5 // indirect
```
Comment (alexec) [inline go.mod:52]: can you run `go mod tidy`?

File: server/auth/sso/sso.go [hunk 3]
```diff
@@ -40,35 +40,79 @@ type sso struct {
 
 type Config struct {
 	Issuer       string                  `json:"issuer"`
-	ClientID     string                  `json:"clientId"`
+	ClientID     apiv1.SecretKeySelector `json:"clientID"`
```
Comment (alexec) [inline server/auth/sso/sso.go:43]: I think this should stay as `clientId`

File: server/auth/sso/sso.go [hunk 4]
```diff
@@ -40,35 +40,79 @@ type sso struct {
...
+// oidc.Provider methods in this file, add them here and provide a stub
+// implementation in test.
+type providerInterface interface {
+	Endpoint() oauth2.Endpoint
+	Verifier(config *oidc.Config) *oidc.IDTokenVerifier
+}
+
+type providerFactory func(ctx context.Context, issuer string) (providerInterface, error)
+
+func providerFactoryOidc(ctx context.Context, issuer string) (providerInterface, error) {
```
Comment (alexec) [inline server/auth/sso/sso.go:59]: should be `providerFactoryOIDC`

File: server/auth/sso/sso.go [hunk 5]
```diff
@@ -40,35 +40,79 @@ type sso struct {
...
+		clientIDObj = clientSecretObj
+	} else {
+		clientIDObj, err = secretsIf.Get(c.ClientID.Name, metav1.GetOptions{})
+		if err != nil {
+			return nil, err
+		}
+	}
+	clientID := clientIDObj.Data[c.ClientID.Key]
+	if clientID == nil {
+		return nil, fmt.Errorf("Key %s missing in secret %s", c.ClientID.Key, c.ClientID.Name)
```
Comment (alexec) [inline server/auth/sso/sso.go:106]: errors should start with lowercase, i.e.` "key $s`

**Comment by alexec** [comment]
@vladlosev I'm online today and I'd love to get this merged today

File: go.mod (hunk 2 above)
Comment (vladlosev) [reply go.mod]: Oops, sorry I didn't mean to commit `go.mod`. But since it's there, got it go-tidied up.

File: server/auth/sso/sso.go (hunk 3 above)
Comment (vladlosev) [reply server/auth/sso/sso.go]: Reverted.

File: server/auth/sso/sso.go (hunk 4 above)
Comment (vladlosev) [reply server/auth/sso/sso.go]: Done.

File: server/auth/sso/sso.go (hunk 5 above)
Comment (vladlosev) [reply server/auth/sso/sso.go]: Done.

**Comment by alexec** [comment]
See #3216

**Comment by vladlosev** [comment]
@alexec you want to merge that one first and have this one rebased?

**Comment by vladlosev** [comment]
Also, updated the PR description with the explanation of the manual testing performed.

File: go.mod [hunk 6]
```diff
@@ -49,6 +49,7 @@ require (
 	github.com/lib/pq v1.3.0 // indirect
 	github.com/mailru/easyjson v0.7.1 // indirect
 	github.com/mattn/go-colorable v0.1.4 // indirect
+	github.com/mattn/go-isatty v0.0.10 // indirect
```
Comment (alexec) [inline go.mod:52]: MIT = ok

**Comment by vladlosev** [comment]
Thanks for helping with this!

**Pull request merged**